src/python/pants/backend/python/framework/stevedore/python_target_dependencies_test.py

```diff
@@ -0,0 +1,233 @@
+                f"runners/{runner}_runner/{runner}_runner/{runner}_runner.py": "",
+                f"runners/{runner}_runner/{runner}_runner/thing.py": "",
+            }
+        )
+
+
+@pytest.fixture
+def rule_runner() -> RuleRunner:
+    rule_runner = RuleRunner(
+        rules=[
+            *python_target_types_rules(),
+            *stevedore_dep_rules(),
+            QueryRule(AllStevedoreExtensionTargets, ()),
+            QueryRule(StevedoreExtensions, ()),
+            QueryRule(InferredDependencies, (InferStevedoreNamespaceDependencies,)),
+        ],
+        target_types=[
+            PythonSourceTarget,
+            PythonSourcesGeneratorTarget,
+            PythonTestTarget,
+            PythonTestsGeneratorTarget,
+            StevedoreExtension,
+        ],
+    )
+    write_test_files(rule_runner)
+    args = [
+        "--source-root-patterns=runners/*_runner",
+    ]
+    rule_runner.set_options(args, env_inherit={"PATH", "PYENV_ROOT", "HOME"})
+    return rule_runner
+
+
+# -----------------------------------------------------------------------------------------------
+# Tests for utility rules
+# -----------------------------------------------------------------------------------------------
+
+
+def test_find_all_stevedore_extension_targets(rule_runner: RuleRunner) -> None:
+    assert rule_runner.request(AllStevedoreExtensionTargets, []) == AllStevedoreExtensionTargets(
+        rule_runner.get_target(Address(f"runners/{runner}_runner", target_name=target_name))
+        for runner in sorted(st2_runners)
+        for target_name in ["runner", "thing"]
+    )
+
+
+def test_map_stevedore_extensions(rule_runner: RuleRunner) -> None:
+    assert rule_runner.request(StevedoreExtensions, []) == StevedoreExtensions(
+        FrozenDict(
+            {
+                "some.thing.else": tuple(
+                    rule_runner.get_target(Address(f"runners/{runner}_runner", target_name="thing"))
+                    for runner in sorted(st2_runners)
+                ),
+                "st2common.runners.runner": tuple(
+                    rule_runner.get_target(
+                        Address(f"runners/{runner}_runner", target_name="runner")
+                    )
+                    for runner in sorted(st2_runners)
+                ),
+            }
+        )
+    )
+
+
+# -----------------------------------------------------------------------------------------------
+# Tests for dependency inference of python targets (python_tests, python_distribution, etc)
+# -----------------------------------------------------------------------------------------------
+
+
+def test_infer_stevedore_namespace_dependencies(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "src/foobar/BUILD": dedent(
+                """\
+                python_tests(
+                    name="tests",
+                    stevedore_namespaces=["st2common.runners.runner"],
```

Comment:
Not easily. StackStorm has a wrappers around stevedore, so you'd have to watch for using certain functions and figure out what the namespace string is. Plus, though we have a constant for it, we actually have 2 copies of it.

So, for the `st2common.runners.runner` namespace, here is where the constant is defined:
https://github.com/StackStorm/st2/blob/master/st2common/st2common/constants/runners.py#L80-L81
but also in this file alongside the some wrapper utils that use it:
https://github.com/StackStorm/st2/blob/master/st2common/st2common/runners/__init__.py
That uses the utils that wrap `stevedore` here:
https://github.com/StackStorm/st2/blob/master/st2common/st2common/util/driver_loader.py
And then, guess what, someone also created nearly the same function here:
https://github.com/StackStorm/st2/blob/master/st2common/st2common/util/loader.py#L210-L225
Plus at least one other place where stevedore itself is used (based on a few quick greps).

And then the question is, how do you know if a **`python_test`** cares about the extensions and expects implementations to be installed? For instance, some of the tests call an rest API to get the list of runners - there is no explicit import or magic string in the API layer that says it needs the runners to be present.

For me, I discover which tests need things through trial and error - try running the test, see if it breaks, and then add the namespaces to fix the errors raised (it will loudly complain if it can't find a requested extension/plugin/driver/... _aside: there are so many names for approximately the same thing. ugh_)

Today, StackStorm's old Makefile just installs all of these extensions/plugins/drivers (eg the runners) before running any of the tests. But, loading all of the runners is a very heavy weight dependency, so I wanted a way to narrow the scope of what needs all of that. This way I'm opting in only specific tests to making those extensions available.